### ISSUE TITLE
0.12.4 png image converted to grayscale

### ISSUE THREAD

**Issue by GeyseR** [wkhtmltopdf/wkhtmltopdf#3387]
Hi!

I wanted to update my wkhtmltopdf installation from 0.12.3 to 0.12.4 version on linux platform (inside docker). Everything works fine except some images become gray after converting to pdf, while in browser they shown correctly with appropriate colors. Colors also works with 0.12.3 version.

Example image: http://2.1m.yt/APdag6O.png
Example html:
```html
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
  </head>
  <body>
    <img src="http://2.1m.yt/APdag6O.png"/>
  </body>
</html>
```

What is wrong with my image and how it can be fixed?

**Comment by alloylab** [comment]
is this still an issue on 0.12.5?

**Comment by GeyseR** [comment]
hi @alloylab. 
Thanks for looking into this issue.
Yes it's still an issue on 0.12.5.
I made some research and found that problem is reproducible when I make opacity for the original image below 20% (i'm using Gimp if it's matter). We have a lot of images which already converted to transparent version and we don't have access to their original images, so we can't just update their opacity to 20%.
Example image from my original post here already expired and unavailable, so i attached new example as zip archive with all sources and result pdf file: 
[wkhtmltopdf-opacity-grayscale-bug.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2162168/wkhtmltopdf-opacity-grayscale-bug.zip)
Could you please take a look?

**Comment by alloylab** [comment]
I was able to reproduce the issue... it seems to be a specific issue regarding img-19.png 
Just to confirm the opacity setting on this image was 20%?

**Comment by GeyseR** [comment]
Right, the issue is reproducible for img-19 
"19" means 19% opacity :) 
i added img-20 (20% opacity) to make it clear that only opacity below 20%  (19%, 18%, 17%....) leads to this issue

**Comment by GeyseR** [comment]
Just tried to save img-19 as pdf in Gimp 2.8, it works perfectly without any changes to colors.
Also, conversion works well on wkhtmltox 0.12.3 versions, so I doubt that there is any limitation in pdf format itself.
My colleague tried to save this file in photoshop, and it works, too
[img-19-photoshop.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2177005/img-19.pdf)
[img-19-gimp.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2177007/img-19.pdf)

**Comment by nook24** [comment]
I also have this issue, after updating from version `0.12.2-1` to `0.12.5-1`.
My PNGs appears to be gray now.

Image in Webbrowser:
![browser](https://user-images.githubusercontent.com/9019992/46201854-7e919600-c315-11e8-8005-3727d9bcbd96.png)

PNG in PDF in `0.12.2-1`:
![0 12 2-1](https://user-images.githubusercontent.com/9019992/46201862-86e9d100-c315-11e8-9eb3-abd80468a3b2.png)

Gray PNG in PDF in `0.12.5-1`:
![0 12 5-1](https://user-images.githubusercontent.com/9019992/46201890-9a953780-c315-11e8-9c96-7729b6fc6af5.png)

Maybe relates to #2208 ? Version `0.13.0-alpha` runs into a core dump on my system so i can't test this.

**Comment by alexchuin** [comment]
We have the same problem here.

**Comment by PhilterPaper** [comment]
@nook24 is there some **transparency** involved here? I think PNG has had trouble with semi-transparent fill. If the colors are totally opaque, there is a problem with the color changing.

**Comment by nook24** [comment]
Yes, the images are using colors like: `rgba(54, 162, 235, 0.2)` (blue) or `rgba(255, 99, 132, 0.2)` (red). ([Source Code](https://github.com/it-novum/openITCOCKPIT/blob/development/app/node/ChartRenderer.js#L24-L31))

It also appears totally random for me, which PNG will be affected by the grayscale. I created two PDFs for you.
One with Version `0.12.2.1`. In this PDF all PNGs works fine.
For the second PDF I used  version `0.12.5-1.bionic` and only the second PNG inside of the PDF is affected. This looks a little bit like a third metric but it is just an rendering issue.
![Bildschirmfoto 2019-05-06 um 16 35 34](https://user-images.githubusercontent.com/9019992/57233295-a6b94400-701e-11e9-8da4-44e68424ec8f.png)
[Test Report-0.12.2.1-1.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/3148464/Test.Report-0.12.2.1-1.pdf)
[Test Report-0.12.5-1.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/3148465/Test.Report-0.12.5-1.pdf)

**Comment by PhilterPaper** [comment]
So what happens if you change all the RGBA transparencies to 1 (fully opaque)? Does the "gray" problem go away? I would think that the last area filled is going to fully cover the earlier areas, which is probably undesirable for your needs. IIRC, the transparency problem may be mixing in _black_ with the rest of the color, rather than _white_. You might double check if black is being used as a default background (see if you can change it to white?). Maybe you can paint the whole background with white, before drawing the data? An alternative might be to use opaque colors, and figure out where you have overlap and explicitly figure out the resulting opaque color, but that's a lot of computational work. Hopefully some of this babbling will give you a clue as to where you can deal with this problem.

**Comment by nook24** [comment]
If I disable transparencies the issue is gone.
Unfortunately the library does not support to set a dedicated background color, so I added everywhere `background: #FFFFFF; background-color: #FFFFFF;`  but this didn't change anything at all.

I noticed that the images that are affected by the issue have a lot of transparent areas that are overlapping each other. (Like the chart itself).
I can generate the PDF as often as i want. It's always the same PNG affected. The image itself looks good in a Browser or image viewer. So i guess it is due to the fact that in this image the metrics in the chart are overlapping a lot more than in other charts.

How ever, the same image just work fine in the older version of wkhtmltopdf:
![wkhtmltopdf](https://user-images.githubusercontent.com/9019992/57276407-c9d90780-70a1-11e9-8251-1144910eb959.png)

**Comment by PhilterPaper** [comment]
I suspect that (unfortunately) it's a bug in WebKit that it doesn't properly mix colors for PNG transparency, perhaps starting with black (0 0 0) rather than white (1 1 1). Whatever exactly is going on, if it's not accessible for change in wkHTMLtoPDF's code, you'll have to wait for WebKit to be updated. That's my guess. Perhaps this will give someone an idea of where to go looking in the code.

**Comment by nook24** [comment]
I'm assuming that the use webkit library gets compiled into wkHTMLtoPDF? That would explain why the rendering behavior is different if I switch from one wkHTMLtoPDF version to another. For now it is ok for me to stay on `0.12.2.1`.

But i will keep an eye on it and test if future versions will resolve the issue...

**Comment by wangjikill** [comment]
Guys, I found such a rule that when loading an image for the first time, it will appear translucent. For example: insert three pictures A, A, B into the page, you will find that the second picture A is normal. Another example: insert pictures A, B, B in the page, you will find that the third picture B is normal. Therefore, I think it is caused by the web engine loading network pictures. When the pictures are cached, it is normal.

**Comment by wangjikill** [comment]
Therefore, if the page is inserted into the network picture, you can insert the same picture twice, and then hide the first picture, and the second picture can be displayed normally. for example:
<img src="http://example.com/a.png" style="visibility:hidden;"/>
<img src="http://example.com/a.png"/>

**Comment by NZKobra** [comment]
I ran into this issue, and it turns out the opacity from a previous DOM element was impacting the PNG rendering.

I had a string of text just before the image, and that text was styled with alpha "color: rgba(0, 0, 0, 0.55)".
Changing the style to remove the alpha channel caused the image to render correctly.

The image src was a base64 data url, but I don't think that matters at all.

I hope this helps.

**Comment by jnnewton** [comment]
I have tried the double loading and it simply moves the issue to a later image that is rendered (which I also double load).   I also did a search of rgba css and I have none in the document.   It appears to be an opacity issue, I can vary the background-color and see the change in the color of the image because it is not opaque.   The issue is never visible in the HTML.   All images are local .JPG, html is local.   From my point of view it appears as though before or during the rendering, wkhtmltopdf is adding an opacity of < 1 to an image.    What is actually happening under the hood is still a mystery and why it only happens to some images (specifically 1 out of 4 in my test case) could make this difficult to track down.
